src/builtin.cpp

```diff
@@ -3028,16 +3028,28 @@ static int builtin_bg(parser_t &parser, io_streams_t &streams, wchar_t **argv) {
             res = send_to_bg(parser, streams, j, _(L"(default)"));
         }
     } else {
-        int i;
-        int pid;
+        std::vector<int> pids;
+
+        // If one argument is not a valid pid (i.e. integer >= 0), fail without backgrounding anything,
+        // but still print errors for all of them.
+        for (int i = 1; argv[i]; i++) {
+            int pid = fish_wcstoi(argv[i]);
+            if (errno || pid < 0) {
+                streams.err.append_format(_(L"%ls: '%ls' is not a valid job specifier\n"), argv[0], argv[i]);
+                res = STATUS_BUILTIN_ERROR;
+            }
+            pids.push_back(pid);
+        }
+        if (res == STATUS_BUILTIN_ERROR) {
+            return res;
+        }
 
-        for (i = 1; argv[i]; i++) {
-            pid = fish_wcstoi(argv[i]);
-            if (errno || pid < 0 || !job_get_from_pid(pid)) {
-                streams.err.append_format(_(L"%ls: '%ls' is not a job\n"), argv[0], argv[i]);
-                return STATUS_BUILTIN_ERROR;
+        // Background all existing jobs that match the pids.
+        // Non-existent jobs aren't an error.
+        for (auto p : pids) {
+            if (job_t* j = job_get_from_pid(p)) {
```

Comment:
I'll be honest: I actually missed that `send_to_bg()` also prints stuff.

So the issue is with "Can't put job to background because it's not under job control". And I'd actually argue for keeping that one, because that case breaks the fundamental assertion that after `bg` every job argument isn't running anymore because it's either stopped or finished.

It's "normal" for bg to encounter a non-existent job because it could have just ended. It's not normal for it to encounter an existing job that it can't handle.

This matters more if `disown` is implemented - if you pass something to `bg` that you just disowned, it's important to know that it won't be handled.

So what we could do, and what you seem to be arguing for, is to print a message if we can't find a job too. I'll implement that.